crates/astria-sequencer/src/mempool/mod.rs

```diff
@@ -149,80 +141,71 @@ impl Mempool {
     #[must_use]
     #[instrument(skip_all)]
     pub(crate) async fn len(&self) -> usize {
-        self.all.read().await.len()
+        #[rustfmt::skip]
+        let (pending_len, parked_len) = join!(
+            async { self.pending.read().await.len() },
+            async { self.parked.read().await.len() }
+        );
+        pending_len.saturating_add(parked_len)
     }
 
     /// Inserts a transaction into the mempool and does not allow for transaction replacement.
     /// Will return the reason for insertion failure if failure occurs.
     #[instrument(skip_all)]
     pub(crate) async fn insert(
         &self,
-        tx: SignedTransaction,
+        tx: Arc<SignedTransaction>,
         current_account_nonce: u32,
     ) -> anyhow::Result<(), InsertionError> {
-        let timemarked_tx = Arc::new(TimemarkedTransaction::new(tx));
-
-        // check if already tracked
-        if self.all.read().await.contains(&timemarked_tx.tx_hash()) {
-            return Err(InsertionError::AlreadyPresent);
-        }
+        let timemarked_tx = TimemarkedTransaction::new(tx);
 
-        // grab needed locks in normal order
-        let mut all = self.all.write().await;
-        let mut pending = self.pending.write().await;
-        let mut parked = self.parked.write().await;
+        let (mut pending, mut parked) = join!(self.pending.write(), self.parked.write());
```

Comment:
Is there a possibility of deadlock here? I'm thinking if two threads calls this (or one of the other holding joins) at the same time, what happens if one thread gets the pending lock and the other the parked lock and then they just both squat on the locks? I tried reading into the `join!` macro more but it didn't seem to say what would happen

Reply:
Good call - it could indeed deadlock!  I'll revert these `join!` calls.